### ISSUE TITLE
Copying WCS instances

### ISSUE THREAD

**Issue by MSeifert04** [astropy/astropy#4989]
I just realized, that copying ``astropy.wcs.WCS`` instances is avaiable using it's ``.copy()`` method.
The documentation already states, that it's just a shallow copy and only to avoid loading the copy module. But  what I missed was that it actually doesn't copy anything what makes a ``WCS`` instance work (``naxis``, ``crpix``, ...). After reading through the source code it became clearer: The important attributes making up the instance are saved as ``Wcsprm`` class and stored as ``.wcs`` attribute. This instance isn't copied when making a shallow copy. And even if that one was copied it would be unlikely that the interesting values are copied because they themself are stored internally as lists.

Maybe it could use some more explanation in the documentation _why_ one would use a shallow copy instead of a deepcopy and what attributes are not copied using ``copy()``. I originally thought about suggesting to remove ``copy()`` altogether in favor of ``deepcopy()`` - but maybe there is some use-case for a shallow copy.

**Comment by mcara** [comment]
> But what I missed was that it actually doesn't copy anything what makes a WCS instance work (naxis, crpix, ...).

What do you mean by this? Do you mean that a copied WCS object does not work properly? Or that you would like a shallow copy to be a behave like a deep copy (i.e., by making a copy of the ``crpix`` array)?

**Comment by MSeifert04** [comment]
@mcara Sorry, I just expected that a normal copy copies the attributes that make up the WCS object: ``crval, crpix, ctype, ...`` because these (in some way) define the ``WCS`` object.

**Comment by mcara** [comment]
If https://github.com/astropy/astropy/issues/4990#issuecomment-222766750 is correct then `WCS.copy()`` was supposed to return a deep copy. However, maybe for each class the developer should decide what kind of copy might be useful to have or allow and thus maybe there should not be a single rule for each class in Astropy. I personally can imagine situations when I would like to have ``.copy()`` to do shallow copy (unlike for ``Header`` class - see #4990 - where I think a shallow copy would be pretty much useless for most users).

**Comment by mcara** [comment]
Example:
```python
>>> from astropy.io import fits
>>> from astropy import wcs
>>> h = fits.open('j8bt06nyq_flt.fits')
>>> w1 = wcs.WCS(h[1].header, h)
>>> h.close()
>>> w2 = w1.copy()
>>> w1.all_pix2world([[1, 2], [3, 4]], 1)
array([[  6.03627933, -72.08835465],
       [  6.03624359, -72.08833574]])
>>> w2.all_pix2world([[1, 2], [3, 4]], 1)
array([[  6.03627933, -72.08835465],
       [  6.03624359, -72.08833574]])
```

Now I ca easily turn off SIP distortions in ``w2`` without *unnecessarily* copying the rest of the object:

```python
>>> w1.sip
<astropy.wcs.Sip object at 0x104cf3260>
>>> w2.sip
<astropy.wcs.Sip object at 0x104cf3260>
>>> w2.sip = None
>>> w1.sip
<astropy.wcs.Sip object at 0x104cf3260>
>>> w2.sip
>>> w1.all_pix2world([[1, 2], [3, 4]], 1)
array([[  6.03627933, -72.08835465],
       [  6.03624359, -72.08833574]])
>>> w2.all_pix2world([[1, 2], [3, 4]], 1)
array([[  6.03628789, -72.08837471],
       [  6.03625212, -72.08835565]])
```

**Comment by MSeifert04** [comment]
@mcara Well I'd expect that a dedicated ``copy`` method would either tell me if some essential attributes are not copied _on purpose_ or that it would copy all these. 

For example I wouldn't expect a ``Table.copy()`` to return references to ``columns`` even though these are just associated instances (and not normally copied with shallow copies).

But if there are real use-cases a small hint in the ``WCS.copy``-documenation of copy would be fine. or just leaving as is. Maybe I'm the only one confused by such things. :smile:

**Comment by mhvk** [comment]
You're definitely not the only one that finds things confusing. Just as a reference point, `Time` has both `copy` method that does copy internal arrays, and a `replicate` method that does not.  These are linked to `__deepcopy__` and `__copy__`, respectively...  `Table.copy` has an option to copy the data or not (linked appropriately to `__deepcopy__` and `__copy__`, resp.)

My own sense is that if one has a `copy` method, it should return an object such that if you change any of its attributes, it does not affect the original. For `Table`, I think the default of copying the columns makes sense since otherwise `copied_table[2] = Row(...)` would overwrite data in the original columns, which would be surprising.  For `Time`, the case is less clear, since the internal arrays should be inmutable, so there is no real reason to copy them.

Anyway, for `WCS`, my sense is that a shallow copy by default is fine as long as "normal usage" on the copy would not change the original. But I don't know to what extent that is the case.

**Comment by mcara** [comment]
I do not want to get into a general discussion on usage of copy vs deepcopy or what other classes/objects do.

However, I just want to mention that it is not clear at all *to me* that internal arrays should be immutable (see @mhvk comment _"For ``Time``, the case is less clear, since the **internal arrays should be inmutable**, so there is no real reason to copy them."_) In connection to WCS, I think attributes that are clearly objects such as (numpy arrays) ``crpix``, ``crval``, ``pc``, (objects) ``sip``, ``wcs``, etc. I expect them to be *mutable* and thus a change to an element of such an array (e.g., ``crpix[0]``) to be "propagated" to the original object _assuming that the change was made on a **shallow copy**_. Almost all attributes of the WCS class are objects (except for ``_naxis1`` and ``_naxis2``) that are mutable and so I expect that if changes to these attributes-objects are made on a shallow copy of a WCS object - these changes should "propagate" to the original WCS object.

I disagree with @mhvk statement:
> _Anyway, for WCS, my sense is that **a shallow copy by default is fine as long as "normal usage" on the copy would not change the original**. But I don't know to what extent that is the case._

I find it to be impossible to reconcile "shallow copy", Python's mutable objects, and "normal usage" on a shallow copy not change the original.

It seems to me that both @MSeifert04 and @mhvk want ``.copy()`` to behave like ``.deepcopy()`` and I do not think this should be the case *specifically for the ``astropy.wcs.WCS`` object*. Both of these functions are clearly documented in the WCS documentation. I do believe that most people should use ``.deepcopy()`` method in most situations. However, unlike ``numpy.array`` whose ``.copy()`` returns a deep copy and it is hard to imagine situations under which one would need shallow copies of the numpy arrays, I can see the use of shallow copies of WCS objects.

Therefore, I do not think there should be any change to the current behavior of ``.copy()`` or ``.deepcopy()`` methods of ``astropy.wcs.WCS`` objects: in my opinion they behave as expected.

**Comment by mcara** [comment]
Maybe maintainers could add a warning to the ``.copy()`` methods such as:

```
.. warning::
    Use deepcopy() instead of copy() unless you know why you need a shallow copy.
```

**Comment by MSeifert04** [comment]
I still think the whole topic of copies is badly defined (maybe by python itself), because there are so many ways to copy classes:

- `copy.copy()`
- `copy.deepcopy()`
- `WCS.copy()`
- `WCS.deepcopy()`
- `WCS[:]` (deepcopy!)

and most classes (not for `WCS`) also implement some `to_*`, `from_*` methods and also allow calling the constructor with an instance of the same class (`WCS(w)` doesn't work fortunatly). And it's really hard to keep track which ones do a deepcopy, which a partially deepcopy and which only do a shallow copy.

Long story short: I like the notice/warning @mcara proposed.

**Comment by astrofrog** [comment]
Note that users in general don't read the docs, so I don't think making ``copy`` be shallow is necessarily a good idea. If anything, it would make more sense to have a single ``copy`` method that defaults to deep copy, and has a ``deep=`` argument that can be set to ``False``. I think our default behavior should reflect what the average user is likely to want.

**Comment by pllim** [comment]
:-1: on shallow copy. If I want them linked, I would just use the `=` sign.

**Comment by mcara** [comment]
@pllim Replace ``w2 = w1.copy()`` in my example in https://github.com/astropy/astropy/issues/4989#issuecomment-222776148 with ``w2 = w1`` as you suggest and see if ``w1.all_pix2world()`` returns same values as in my example.

I do not understand why you do not like shallow copies but they can be very useful and shallow copies are quite useful when dealing with WCS objects.

**Comment by mcara** [comment]
@astrofrog I think you idea of having a single copy method with deep (or shallow) parameter is quite interesting. Unfortunately, I think WCS has ``.copy()`` and ``.deepcopy()``since its inception as PyWCS (I did not verify this statement) and changing the behavior of ``.copy()`` or removing ``.deepcopy()`` will affect a lot of code out there.

At the same time I do not think there is much one can do about users who do not read documentation. If users do not read documentation I will be happy to not document my methods in the future (why waste time?)

**Comment by MSeifert04** [comment]
I just realized I haven't provided the example that made me wonder, it's more or less:

```
>>> from astropy.wcs import WCS
>>> w1 = WCS(naxis=2)
>>> w2 = w1.copy()
>>> w2.wcs.crpix = (1,1)  # new crpix.
>>> w1
WCS Keywords

Number of WCS axes: 2
CTYPE : ''  ''  
CRVAL : 0.0  0.0  
CRPIX : 1.0  1.0          # oups
PC1_1 PC1_2  : 1.0  0.0  
PC2_1 PC2_2  : 0.0  1.0  
CDELT : 1.0  1.0  
NAXIS    : 0 0
```
Even if I'd prefer an approach that would partially or totally deepcopy here and adjust ``pix2world`` to optionally disable particular corrections - I understand the reasons why one doesn't need to change anything here.

Except for the note/warning in the copy method. That should be included for those who read documentation :confused: .

**Comment by pllim** [comment]
Can we close this now?

**Comment by MSeifert04** [comment]
Yes!
